### ISSUE TITLE
Macro hygene and generics

### ISSUE THREAD

**Issue by s3bk** [rust-lang/rust#42895]
Identifies declared in macros collide with identifiers passed as macro variables.

I would expect that the following code macro works for with any identifier [(gist)](https://gist.github.com/8849928cdee1f15b2464b35d15b3acfc):

```rust
trait X<A> {}
struct Y<A, B>(A, B);
struct Z<A, B>(A, B);

macro_rules! x {
    ($S:ident, $A:ident) => (
        impl<$A, B> X<B> for $S<$A, B> {}
    )
}

x!(Y, A); // works
x!(Z, B); // collides
```
But it results in: 
```
error[E0403]: the name `B` is already used for a type parameter in this type parameter list
```

**Comment by petrochenkov** [comment]
Hygiene doesn't apply to type parameters in `macro_rules!`. This is a long standing issue that can't be fixed backward compatibly in `macro_rules!`, but it is fixed in (currently unstable) macros 2.0:
```rust
#![feature(decl_macro)]
#![allow(unused)]

trait X<A> {}
struct Y<A, B>(A, B);
struct Z<A, B>(A, B);

macro x($S:ident, $A:ident) {
    impl<$A, B> X<B> for $S<$A, B> {}
}

x!(Y, A); // works
x!(Z, B); // works

fn main() {}
```

**Comment by s3bk** [comment]
Thank you! that explains it.
